storage/test/sparse_merkle/tree_test.cpp

```diff
@@ -273,13 +406,658 @@ TEST(tree_tests, multiple_updates) {
+// remove should return a stale root node and an updated root node.
+TEST(tree_tests, remove_complete_at_root_test) {
+  std::shared_ptr<TestDB> db(new TestDB);
+  Tree tree(db);
+  SetOfKeyValuePairs updates;
+  updates.emplace(Sliver("key1"), Sliver("val1"));
+  updates.emplace(Sliver("key2"), Sliver("val2"));
+  auto batch = tree.update(updates);
+  // There should be two leafs in the root
+  ASSERT_EQ(2, batch.internal_nodes[0].second.numLeafChildren());
+  db_put(db, batch);
+
+  // Delete one of the children
+  KeysVector deletes{Sliver("key1")};
+  batch = tree.remove(deletes);
+  db_put(db, batch);
+
+  // The root should only have one leaf child, and should be the only internal
+  // node that needs writing.
+  ASSERT_EQ(1, batch.internal_nodes.size());
+  ASSERT_EQ(0, batch.leaf_nodes.size());
+  auto& [root_key, root_node] = batch.internal_nodes[0];
+  ASSERT_EQ(Version(2), root_key.version());
+  ASSERT_EQ(Version(2), root_node.version());
+  ASSERT_EQ(1, root_node.numLeafChildren());
+
+  // The old root node is the only stale internal node. The deleted LeafKey at
+  // the version 1 should match the hash of "key1".
+  ASSERT_EQ(Version(2), batch.stale.stale_since_version);
+  ASSERT_EQ(1, batch.stale.internal_keys.size());
+  ASSERT_TRUE(internalKeyExists("", 1, batch.stale.internal_keys));
+  ASSERT_EQ(1, batch.stale.leaf_keys.size());
+  ASSERT_TRUE(leafKeyExists("key1", 1, batch.stale.leaf_keys));
+}
+
+// Insert 2 keys that match in the first Nibble, and are inserted at the same
+// InternalNode at depth 1. Deleting one of them should cause the other one to
+// be promoted, and reside in the root. The root is the only new InternalNodeKey
+// to be written.
+TEST(tree_tests, remove_and_promote) {
+  std::shared_ptr<TestDB> db(new TestDB);
+  Tree tree(db);
+  SetOfKeyValuePairs updates;
+  updates.emplace(Sliver("key3"), Sliver("val1"));
+  updates.emplace(Sliver("key9"), Sliver("val2"));
+  auto batch = tree.update(updates);
+
+  // There should be 2 internal nodes, with no leaf children in the root node.
+  ASSERT_EQ(2, batch.internal_nodes.size());
+  auto& root_node = batch.internal_nodes[0].second;
+  ASSERT_EQ(0, root_node.numLeafChildren());
+  db_put(db, batch);
+
+  // Delete one of these nodes and cause the other to be promoted.
+  KeysVector deletes{Sliver("key9")};
+  batch = tree.remove(deletes);
+
+  // Only the root node exists in the tree now, and it's at version 2.
+  ASSERT_EQ(1, batch.internal_nodes.size());
+  auto& [root_key, remove_root_node] = batch.internal_nodes[0];
+  ASSERT_EQ("", root_key.path().toString());
+  ASSERT_EQ(Version(2), root_key.version());
+  ASSERT_EQ(Version(2), remove_root_node.version());
+
+  // There are no leaf updates, since key3 was not updated. A pointer
+  // (LeafChild) was just moved to the root node.
+  ASSERT_EQ(0, batch.leaf_nodes.size());
+  ASSERT_EQ(1, remove_root_node.numLeafChildren());
+  ASSERT_TRUE(leafChildExists("key3", 1, remove_root_node));
+
```

Comment:
Should we check `batch.stale.leaf_keys` for `key9` here?

Reply:
Done. I also made it so that we check that proper internal node is removed as well.